### PR TITLE
Don't assume the return is in the last block

### PR DESCRIPTION
**Pull request by s-perron**

### CODE REVIEW

File: lgc/patch/FragColorExport.cpp [hunk 1]
```diff
@@ -502,7 +502,17 @@ bool LowerFragColorExport::runOnModule(Module &module) {
...
-  auto retInst = fragEntryPoint->back().getTerminator();
+  auto &bbList = fragEntryPoint->getBasicBlockList();
+  ReturnInst *retInst = nullptr;
+  for (auto bb = bbList.rbegin(); bb != bbList.rend(); ++bb) {
+    auto *terminator = bb->getTerminator();
+    if (isa<ReturnInst>(terminator)) {
+      retInst = dyn_cast<ReturnInst>(terminator);
+      break;
+    }
+  }
```
Comment (kuhar) [inline lgc/patch/FragColorExport.cpp:513]: I think more idiomatic code would look something like this:

```
ReturnInst *retInst = nullptr;
for (auto &block : llvm::reverse(*fragEntryPoint)) {
  if (auto ret = dyn_cast<ReturnInst>(block.getTerminator())) {
    retInst = ret;
    break;
  }
}
```

Comment (s-perron) [reply lgc/patch/FragColorExport.cpp]: Done

Comment (s-perron) [reply lgc/patch/FragColorExport.cpp]: Oh, I missed that part.  That is nice.

Comment (kuhar) [reply lgc/patch/FragColorExport.cpp]: Thanks

**Comment by amdvlk-admin** [comment]
### Test summary for commit 77abb420cbd9a85f19ecbdf91b01027ee51de2e2
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 712f76fee75d69b23a1ea8f6465752c3ccaaf9a2</li>
<li>PAL: dev ea5db60841dab7d067f5010f28a980ef222bdf81</li>
<li>SPVGEN: dev 8dc855026f2502ab3f45dadaf0bb802a57d6ad60</li>
<li>XGL: dev 8d9da0426ba869f5f3da23fd72cfaa669ce99c25</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 8abd9486af1727c8c38afcfef7390d70bb7d7d89</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/441050)</summary><ul><li>Built with branch Vulkan-CTS-1.2.0.2-fetch-sources-fix</li>
<details open><summary>Rhel 8.2, Gfx8</summary><ul>

<li>Passed: 146974/146974 (100.0%)</li>
<li>Failed: 0/146974 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx10</summary><ul>

<li>Passed: 147018/147018 (100.0%)</li>
<li>Failed: 0/147018 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 147058/147058 (100.0%)</li>
<li>Failed: 0/147058 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/FragColorExport.cpp [hunk 2]
```diff
@@ -502,7 +502,19 @@ bool LowerFragColorExport::runOnModule(Module &module) {
   if (!fragEntryPoint)
     return false;
 
-  auto retInst = fragEntryPoint->back().getTerminator();
+  // Find the return instruction as that will be the insertion point for the export instructions.
+  // It is possible that there is no return instruction if there is an infinite loop.  See the shaderdb test
+  // OpLoopMerge_TestIterationControls_lit.frag.  In that case, there should be no need for exports.
+  ReturnInst *retInst = nullptr;
+  for (auto &bb : llvm::reverse(*fragEntryPoint)) {
```
Comment (amdrexu) [inline lgc/patch/FragColorExport.cpp:509]: bb -> block

Comment (s-perron) [reply lgc/patch/FragColorExport.cpp]: Done.

**Pull request merged**